enhancements/storage/csi-cinder-operator.md

```diff
@@ -0,0 +1,179 @@
+
+## Summary
+
+In the past storage drivers have been provided either in-tree or via external provisioners. With the move to advocating CSI as the way to deliver storage drivers, OpenShift needs to move the in-tree drivers (and others that our customers need) to use CSI. With the focus on operators in Red Hat OCP 4.x these CSI drivers should be installed in the form of an Operator.
+
+## Motivation
+
+- In-tree drivers will be removed from Kubernetes, we need to continue to make the drivers available and CSI is the way to do this.
+
+- OpenStack is a key cloud provider for OpenShift and use of Cinder has been supported in both OpenShift 3.x and 4.x, we need to move this driver to be CSI, provided by an operator.
+
+- CSI drivers (including OpenStack Cinder) will enable new storage features, not possible with in-tree volume plugins, like snapshots and volume cloning.
+
+### Goals
+
+- Create a cluster operator to install the OpenStack Cinder CSI driver.
+
+- Publish the operator on OperatorHub.
+
+- Package and ship downstream images for the operator and OpenStack Cinder driver.
+
+### Non-Goals
+
+- Driver creation as it is available in [upstream](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-csi-plugin.md).
+
+- Implement CSI tooling (snapshotter, resizer, cloner). It is available in OpenShift since 4.3.
+
+## Proposal
+
+OCP ships a new cluster operator called cinder-csi-driver-operator.
+
+1. The operator installs the OpenStack Cinder CSI driver following the [recommended guidelines (currently work-in-progress)](https://github.com/openshift/enhancements/pull/139/files).
+2. The operator deploys the all the objects required by the OpenStack Cinder CSI driver:
+   2.1 A namespace called `openshift-cinder-csi-driver`.
+   2.2 Two ServiceAccounts: one for the Controller Service and other for the Node Service of the CSI driver.
+   2.3 The RBAC rules to be used by the sidecar containers.
+   2.4 The CSIDriver object representing the CSI driver.
+   2.5 A Deployment that runs the driver's Controller Service.
+   2.7 A DaemonSet that runs the driver's Node Service.
+   2.8 A non-default StorageClass that uses the CSI driver as a provisioner.
+3. The operator deploys all the CSI driver objects in the namespace `openshift-cinder-csi-driver`.
+   3.1 This is true regardless the namespace where the operator itself is deployed.
+4. The operator itself is installed by [Cluster Storage Operator](https://github.com/openshift/cluster-storage-operator) for all OpenStack clouds.
+   4.1 The operator should be deployed in the namespace `openshift-cluster-csi-drivers`.
+   4.2 The CR which the operator reacts to is *non-namespaced* and is named `cinder.csi.openstack.org`.
+   4.3 The CR belongs to `ClusterCSIDriver` CRD.
+5. The operator leverages the OpenStack credentials by creating a credentials request to [Cluster Credential Operator](https://github.com/openshift/cloud-credential-operator).
```

Comment:
Should there be a note about self-signed OpenStack certificates?

Reply:
I don't think so. Unlike Manila, Cinder operator doesn't communicate with the underlying OpenStack cloud, so SSC support is not required. For the driver itself we just fetch certs from the config map and mount to the pods as usual.

Reply:
Still, it must be tested by QA.